FWCore/Framework/interface/InputProcessBlockCacheImpl.h

```diff
@@ -0,0 +1,234 @@
+        // the branch is present and this index is valid, the product might not be there.
+        // The functor in the CacheFiller must deal with that case.
+        if (index != ProcessBlockHelperBase::invalidCacheIndex()) {
+          std::get<I>(cacheHandles) = CacheHandle(std::get<I>(caches_.at(index).cacheTuple_).get());
+        }
+        fillTuple<I + 1>(cacheHandles, event);
+      }
+
+      std::tuple<CacheHandle<CacheTypes>...> processBlockCaches(Event const& event) const {
+        std::tuple<CacheHandle<CacheTypes>...> cacheHandles;
+        // processNames will be empty if and only if registerProcessBlockCacheFiller
+        // was never called by the module constructor
+        if (!processNames_.empty()) {
+          fillTuple<0>(cacheHandles, event);
+        }
+        return cacheHandles;
+      }
+
+      void selectInputProcessBlocks(ProductRegistry const& productRegistry,
+                                    ProcessBlockHelperBase const& processBlockHelperBase,
+                                    EDConsumerBase const& edConsumerBase) {
+        unsigned int i = 0;
+        for (auto const& tokenInfo : tokenInfos_) {
+          if (!tokenInfo.token_.isUninitialized()) {
+            ProductLabels productLabels;
+            edConsumerBase.labelsForToken(tokenInfo.token_, productLabels);
+
+            processNames_[i] = processBlockHelperBase.selectProcess(productRegistry, productLabels, tokenInfo.typeID_);
+          }
+          ++i;
+        }
+        tokenInfos_.clear();
+      }
+
+      template <std::size_t N>
+      using CacheTypeT = typename std::tuple_element<N, std::tuple<CacheTypes...>>::type;
+
+      template <std::size_t ICacheType, typename DataType, typename Func>
+      void registerProcessBlockCacheFiller(EDGetTokenT<DataType> const& token, Func&& func) {
+        registerProcessBlockCacheFiller<ICacheType, CacheTypeT<ICacheType>, DataType, Func>(token,
+                                                                                            std::forward<Func>(func));
+      }
+
+      template <typename CacheType, typename DataType, typename Func>
+      void registerProcessBlockCacheFiller(EDGetTokenT<DataType> const& token, Func&& func) {
+        static_assert(countTypeInParameterPack<CacheType, CacheTypes...>() == 1u,
+                      "If registerProcessBlockCacheFiller is called with a type template parameter\n"
+                      "then that type must appear exactly once in the template parameters of InputProcessBlockCache");
+
+        // Find the index into the parameter pack from the CacheType
+        constexpr unsigned int I = indexInputProcessBlockCache<CacheType, CacheTypes...>();
+
+        registerProcessBlockCacheFiller<I, CacheType, DataType, Func>(token, std::forward<Func>(func));
+      }
+
+      // This gets used for stream type modules instead of registerProcessBlockCacheFiller
+      void copyProcessBlockCacheFiller(std::vector<edm::impl::TokenInfo>& tokenInfos,
```

Comment:
Given that the arguments are moved from, would `moveProcessBlockCacheFiller()` be a better name?

Reply:
Done, I agree that name is better.